lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -552,6 +568,272 @@ class ConvertAtenMaxPool2dWithIndicesBackwardOp
+                  Value ind = b.create<linalg::IndexOp>(loc, 0);
+                  for (int i = indexType.getRank() - 1; i >= 0; i--) {
+                    indices[i] =
+                        b.create<arith::RemSIOp>(loc, ind, indexShape[i]);
+                    ind = b.create<arith::DivSIOp>(loc, ind, indexShape[i]);
+                  }
+                  // Extract values
+                  Value extractIndexValue =
+                      b.create<tensor::ExtractOp>(loc, adaptor.getIndex(), indices);
+                  Value extractSrcValue =
+                      b.create<tensor::ExtractOp>(loc, adaptor.getSrc(), indices);
+                      SmallVector<Value> yieldVals;
+                  for (Value v : indices) {
+                    Value scalar = castIndexToInt64(b, loc, v);
+                    yieldVals.push_back(b.create<arith::TruncIOp>(
+                        loc, rewriter.getI32Type(), scalar));
+                  }
+                  yieldVals[dim] = b.create<arith::TruncIOp>(
+                      loc, rewriter.getI32Type(), extractIndexValue);
+                  yieldVals.push_back(extractSrcValue);
+                  b.create<linalg::YieldOp>(loc, yieldVals);
+    }).getResultTensors();
+
+    auto toOpFoldResult = [](Value v) -> OpFoldResult {
+      auto op = v.getDefiningOp<arith::ConstantIndexOp>();
+      if (!op)
+        return v;
+      return op.getValue();
+    };
+
+    // concat index tensors
+    SmallVector<Value> offsets, strides;
+    strides.resize(2, rewriter.create<arith::ConstantIndexOp>(loc, 1));
+    offsets.resize(2, rewriter.create<arith::ConstantIndexOp>(loc, 0));
+    Value cstIndexRank =
+        rewriter.create<arith::ConstantIndexOp>(loc, indexType.getRank());
+    Value indices = createZeroInitTensor(
+        rewriter, loc, SmallVector<Value>({indexSize, cstIndexRank}),
+        rewriter.getI32Type());
+    SmallVector<Value> scatterInputsVector(tmTensorScatterInputs);
+    for (auto slice : ArrayRef(scatterInputsVector).drop_back()) {
+      SmallVector<Value> sizes = getTensorSizes(rewriter, loc, slice);
+      indices = rewriter.createOrFold<tensor::InsertSliceOp>(
+          loc, slice, indices,
+          llvm::to_vector(llvm::map_range(offsets, toOpFoldResult)),
+          llvm::to_vector(llvm::map_range(sizes, toOpFoldResult)),
+          llvm::to_vector(llvm::map_range(strides, toOpFoldResult)));
+      offsets[1] =
+          rewriter.createOrFold<arith::AddIOp>(loc, offsets[1], cstOne);
+    }
+    // Cast to static shapes
+    SmallVector<int64_t> staticShape({ShapedType::kDynamic, indexType.getRank()});
+
+    auto staticType = RankedTensorType::get(staticShape, rewriter.getI32Type());
+    indices = rewriter.create<tensor::CastOp>(loc, staticType, indices);
+
+    Value updates = scatterInputsVector[indexType.getRank()];
+
+    // @TODO: Create special count tensor to calculate the 'mean'
+    Value counts = nullptr;
+    if (reduceType == "mean") {
```

Comment:
Can you copy

https://github.com/pytorch/pytorch/blob/0a9c608461f53a6c7dc019199f6c93860efb4c0b/aten/src/ATen/native/ReductionType.h#L7-L23

into

https://github.com/llvm/torch-mlir/blob/main/include/torch-mlir/Dialect/Torch/Utils/TorchUpstream.h

and use the enum instead of the strings?

Reply:
Done